externals/llvm-external-projects/torch-mlir-dialects/lib/Dialect/TMTensor/IR/TMTensorOps.cpp

```diff
@@ -84,6 +84,254 @@ OpFoldResult TMTensor::getDim(OpBuilder &builder, Location loc, Value v,
+                   x = b.create<arith::AddFOp>(loc, x, acc);
+
+                   b.create<scf::YieldOp>(loc, x);
+                 })
+                ->getResult(0);
+        b.create<memref::StoreOp>(loc, sum, output, localIVs);
+        b.create<scf::YieldOp>(loc);
+      });
+}
+
+LogicalResult AttentionOp::generateScalarImplementation(OpBuilder &b,
+                                                        Location loc,
+                                                        ValueRange ivs) {
+
+  Value query = getQuery();
+  Value key = getKey();
+  Value value = getValue();
+  Value output = getOutput();
+  auto queryType = query.getType().cast<MemRefType>();
+  auto keyType = key.getType().cast<MemRefType>();
+  auto valueType = value.getType().cast<MemRefType>();
+  auto queryRank = queryType.getRank();
+  auto keyRank = keyType.getRank();
+  auto valueRank = valueType.getRank();
+  auto keySizes = keyType.getShape();
+  Type elementType = queryType.getElementType();
+
+  Value zeroF = b.create<arith::ConstantOp>(loc, elementType,
+                                            b.getFloatAttr(elementType, 0.0));
+
+  SmallVector<Value> queryDynSizes, keyDynSizes, valueDynSizes, outputDynSizes;
+  for (auto i = 0; i < queryRank; i++)
+    queryDynSizes.push_back(b.create<memref::DimOp>(loc, query, i));
+  for (auto i = 0; i < keyRank; i++)
+    keyDynSizes.push_back(b.create<memref::DimOp>(loc, key, i));
+  for (auto i = 0; i < valueRank; i++)
+    valueDynSizes.push_back(b.create<memref::DimOp>(loc, value, i));
+  for (auto i = 0; i < queryRank; i++)
+    outputDynSizes.push_back(b.create<memref::DimOp>(loc, output, i));
+
+  // weight = query @ key
+  auto weightSizes = SmallVector<int64_t>(queryType.getShape());
+  weightSizes[queryRank - 1] = keySizes[keyRank - 1];
+  auto weightType = MemRefType::get(weightSizes, queryType.getElementType());
+  SmallVector<Value> weightDynSizes(queryDynSizes);
+  weightDynSizes[weightDynSizes.size() - 1] =
+      keyDynSizes[keyDynSizes.size() - 1];
+  Value weight = b.create<memref::AllocOp>(loc, weightType, weightDynSizes);
+  matmul(b, loc, query, queryDynSizes, key, keyDynSizes, weight, weightDynSizes,
+         true);
+
+  // weight = softmax(weight)
+  Value one = b.create<arith::ConstantIndexOp>(loc, 1);
+  Value zero = b.create<arith::ConstantIndexOp>(loc, 0);
+  Value dim = weightDynSizes[weightDynSizes.size() - 1];
+  Value scaleFactor = b.create<math::SqrtOp>(
+      loc, b.create<arith::UIToFPOp>(
+               loc, elementType,
+               b.create<arith::IndexCastUIOp>(loc, b.getI32Type(),
+                                              queryDynSizes[queryRank - 1])));
+  // calculate max(weight)
+  Value init = b.create<memref::LoadOp>(
+      loc, weight, SmallVector<Value>(weightDynSizes.size(), zero));
+  Value globalMax =
+      b.create<scf::ParallelOp>(
+           loc, SmallVector<Value>(queryRank, zero), weightDynSizes,
+           SmallVector<Value>(queryRank, one), init,
```

Comment:
There are a few instances in this pattern where the `weight` tensor is being iterated over, but the bounds are specified in terms of `queryRank`. I understand that `queryRank == weightRank`, but I think it would be easier to read if a variable `weightRank` was used instead. This makes it very clear that it is just an iteration over the `weight`